Port MeshPipe to use GCDAsyncUdpSocket for better handling and performance

bogo opened the pull request:
MeshPipe currently uses the vanilla, run loop-based flavor of [CocoAsyncSocket](https://github.com/robbiehanson/CocoaAsyncSocket), the `AsyncUdpSocket`. Since its `-doReceive` is a blocking call, this will cause performance issues and lock ups in applications using MeshPipe.

This patch:
- removes the foreign code of `CocoaAsyncSocket`
- removes the foreign code of `GZIP`
- adds the `CocoaAsyncSocket` submodule
- adds the `GZIP` submodule
- ports MeshPipe to use the vanilla `GCDAsyncUdpSocket`
- adds the IPv6 loopback (`::1`) to the list of local hosts checked

This was tested with [CoreDragon](https://github.com/nevyn/CoreDragon) and seems to both improve performance and retain all existing functionality.

cc @nevyn.

bogo committed:
Port MeshPipe to use GCDAsyncUdpSocket

nevyn commented on MeshPipe/MeshPipe.m:
```diff
@@ -7,25 +7,30 @@
+static NSString *MeshPipeLocalAddress = @"127.0.0.1";
 
-//#define MPDebug(...) NSLog(__VA_ARGS__)
 #define MPDebug(...)
```
This should yield a "redefinition" warning in Release builds; put it in an #elif plz

bogo replied:
ah, of course!

nevyn commented on MeshPipe/MeshPipe.m:
```diff
@@ -64,36 +80,39 @@ - (instancetype)initWithBasePort:(int)basePort count:(int)count peerName:(NSStri
 		int port = _basePort + i;
-		BOOL success = [_listenSocket bindToAddress:@"localhost" port:port error:NULL];
+        NSData *address = [MeshPipe _addressFor:MeshPipeLocalAddress port:port];
+		BOOL success = [_listenSocket bindToAddress:address error:NULL];
```
Why this dance instead of "localhost"? wanna add an explanatory comment?

bogo replied:
oh, `GCDAsyncSocket` doesn't have the idea of binding to a string-based address; you have to pass it a data structure built off `sockaddr_in`.

bogo replied:
It can bind to port though. And it does allow to connect to string/port combination. So either I can bind all of them to port + string/port, or do `NSData` addresses everywhere. (for sake of coherence)

nevyn commented on MeshPipe/MeshPipe.m:
```diff
@@ -64,36 +80,39 @@ - (instancetype)initWithBasePort:(int)basePort count:(int)count peerName:(NSStri
 	}
 	if(!_myPort) {
-		NSLog(@"No free MeshPipe slots, not creating pipe");
+		MPDebug(@"No free MeshPipe slots, not creating pipe");
```
imo fatal error should print to console; do you disagree?

bogo replied:
I agree that fatal errors should go to console. I don't have a strong opinion whether `MPLogError` or `NSLog` would be preferred here. I'll add `MPLogError`. :)

nevyn commented on MeshPipe/MeshPipe.m:
```diff
@@ -64,36 +80,39 @@ - (instancetype)initWithBasePort:(int)basePort count:(int)count peerName:(NSStri
-			NSLog(@"Unexectedly couldn't connect to MeshPipe potential peer %d, giving up: %@", i, err);
+        NSData *address = [MeshPipe _addressFor:MeshPipeLocalAddress port:potentialPeer.port];
+		if(![potentialPeer.socket connectToAddress:address error:&err]) {
+			MPDebug(@"Unexectedly couldn't connect to MeshPipe potential peer %d, giving up: %@", i, err);
```
fatal error thingie

nevyn commented:
@bogo thanksss <3 added some comments. could fix them myself but iz too tired rn

nevyn commented:
also probably need to update podfile to make the dependencies work out....... ugh

bogo committed:
Use the GZIP submodule instead of foreign code

bogo committed:
Add .gitignore for Objective-C

bogo committed:
Add CocoaAsyncSocket to the Xcode project

bogo committed:
Add the new dependencies to Podfile

bogo committed:
MPDebug to MPLogDebug and MPLogError

bogo committed:
Bind directly to localhost, instead of using NSData-based addresses

bogo commented:
I changed things according to your comments. I also removed the `NSData` dance; seems everything works fine just by directly binding to a port. Plus, I adjusted the Xcode project file and removed the GZIP code. Everything still works in my test app. ;)

I'm not sure if I updated the Pods file correctly, so if you can give it a look over, that'd be fantastic!

nevyn commented:
well that's pretty awesome. Thanks so much for this @bogo master!

Not sure if the submos will cause problems for others, but I think it should be fine. let's try it.
